### PR TITLE
Navigation Integration

### PR DESCRIPTION
**Pull request by electrostat**
- update events and metrics to match v1 navigation events

### CODE REVIEW

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationUtils.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    int screenBrightness;
+    try {
+      screenBrightness = android.provider.Settings.System.getInt(
+        MapboxTelemetry.applicationContext.getContentResolver(),
+        android.provider.Settings.System.SCREEN_BRIGHTNESS);
+
+      // Android returns values between 0 and 255, here we normalize to 0-100.
+      screenBrightness = (int) Math.floor(100.0 * screenBrightness / 255.0);
+    } catch (Settings.SettingNotFoundException exception) {
+      screenBrightness = -1;
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationUtils.java]: Magic number.

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationUtils.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    return screenBrightness;
+  }
+
+  static String obtainAudioType() {
+    AudioTypeChain audioTypeChain = new AudioTypeChain();
+    AudioTypeResolver setupChain = audioTypeChain.setup();
+
+    return setupChain.obtainAudioType(MapboxTelemetry.applicationContext, audioManager);
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationUtils.java]: 👀 previous comment on `AudioManager`

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 3]
```diff
@@ -21,6 +26,118 @@
   private static final String THREE_STRING_FORMAT = "%s/%s/%s";
   private static final SimpleDateFormat dateFormat = new SimpleDateFormat(DATE_AND_TIME_PATTERN, Locale.US);
   private static final Locale DEFAULT_LOCALE = Locale.US;
+  private static final int UNAVAILABLE_BATTERY_LEVEL = 100;
+  static TelephonyManager telephonyManager = null;
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: Why do we need a `static` `TelephonyManager` variable? With a `Context` you can obtain a `TelephonyManager` as was implemented before in `MapEventFactory#obtainCellularNetworkType` 
```java
TelephonyManager manager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
```
so declaring a `static` variable here won't be necessary.

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 4]
```diff
@@ -21,6 +26,118 @@
...
+  static ActivityManager activityManager = null;
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: Why do we need a `static` `ActivityManager` variable? With a `Context` you can obtain a `ActivityManager` so declaring a `static` variable here won't be necessary.

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 5]
```diff
@@ -21,6 +26,118 @@
...
+      for (int j = i; j < length; j += Character.charCount(c)) {
+        c = s.codePointAt(j);
+        buffer.writeUtf8CodePoint(c > '\u001f' && c < '\u007f' ? c : '?');
+      }
+      return buffer.readUtf8();
+    }
+    return s;
+  }
+
+  static String getApplicationState() {
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: `get` nomenclature is kind of _kidnapped_ by Java and it's usually related with getters of a POJO. What about using for example `obtain` instead and give it more semantics? The same thing happens with `getBatteryLevel()` and `getCellularNetworkType()`.

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 6]
```diff
@@ -21,6 +26,118 @@
...
+    if (activityManager == null) {
+      activityManager = (ActivityManager) MapboxTelemetry.applicationContext.getSystemService(Context.ACTIVITY_SERVICE);
+    }
+
+    List<ActivityManager.RunningAppProcessInfo> appProcesses = activityManager.getRunningAppProcesses();
+    if (appProcesses == null) {
+      return "";
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: Magic number.

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 7]
```diff
@@ -21,6 +26,118 @@
...
+    }
+
+    String packageName = MapboxTelemetry.applicationContext.getPackageName();
+    for (ActivityManager.RunningAppProcessInfo appProcess : appProcesses) {
+      if (appProcess.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND
+        && appProcess.processName.equals(packageName)) {
+        return "Foreground";
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: Magic number.

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 8]
```diff
@@ -21,6 +26,118 @@
...
+      }
+    }
+
+    return "Background";
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: Magic number.

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 9]
```diff
@@ -21,6 +26,118 @@
...
+  }
+
+  static int getBatteryLevel() {
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: Why not using the already refactored and discussed code from `MapEventFactory#obtainBatteryLevel` (removed as part of this PR)?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 10]
```diff
@@ -21,6 +26,118 @@
...
+    if (batteryStatus != null) {
+      int level = batteryStatus.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
+      int scale = batteryStatus.getIntExtra(BatteryManager.EXTRA_SCALE, -1);
+      return Math.round((level / (float) scale) * 100);
+    } else {
+      return UNAVAILABLE_BATTERY_LEVEL;
+    }
+  }
+
+  static boolean isPluggedIn() {
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: Why not using the already refactored and discussed code from `MapEventFactory#isPluggedIn` (removed as part of this PR)?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 11]
```diff
@@ -21,6 +26,118 @@
...
+      int chargePlug = batteryStatus.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1);
+      if (chargePlug == BatteryManager.BATTERY_PLUGGED_USB
+        || chargePlug == BatteryManager.BATTERY_PLUGGED_AC) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  static String getCellularNetworkType() {
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: Why not using the already refactored and discussed code from `MapEventFactory#obtainCellularNetworkType` (removed as part of this PR)?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/AudioTypeResolver.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package com.mapbox.android.telemetry;
+
+import android.content.Context;
+import android.media.AudioManager;
+
+interface AudioTypeResolver {
+  void nextChain(AudioTypeResolver chain);
+
+  String obtainAudioType(Context context, AudioManager audioManager);
```
[the remark this answers is not in the document]
Comment (electrostat) [reply libtelemetry/src/main/java/com/mapbox/android/telemetry/AudioTypeResolver.java]: This solution was created to pass tests. Unit tests were failing when mocking and passing along context alone, so reworked code to pass along AudioManager and a few other system services that required context. This allowed me to mock those services and pass tests, while still including desired functionality and refactoring.

[the remark this answers is not in the document]
Comment (Guardiola31337) [reply libtelemetry/src/main/java/com/mapbox/android/telemetry/AudioTypeResolver.java]: We should avoid adding production code just for testing.
In this case, we can fix the tests using a stub for `AudioManager`.
We're already implementing that technique in many other tests. Here there is an example with `TelephonyManager` 👀 https://github.com/mapbox/mapbox-events-android/blob/7fd4005be44943a2f2649cf5c7fe2abb6503dbb8/libtelemetry/src/test/java/com/mapbox/android/telemetry/MapEventFactoryTest.java#L133-L135

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationCancelData.java [hunk 13]
```diff
@@ -4,11 +4,13 @@
 import android.os.Parcelable;
 
 public class NavigationCancelData implements Parcelable {
-  private String arrivalTimestamp;
+  private String arrivalTimestamp = null;
   private Integer rating = null;
   private String comment = null;
 
-  public NavigationCancelData(String arrivalTimestamp) {
+  public NavigationCancelData() {}
```
[the remark this answers is not in the document]
Comment (electrostat) [reply libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationCancelData.java]: System will default to using the parcel method if a separate method is not provided.

https://github.com/mapbox/mapbox-events-android/blob/69994b636e604e5b6fcde4d0daa2d6365ad75931/libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationCancelEvent.java#L39

[the remark this answers is not in the document]
Comment (Guardiola31337) [reply libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationCancelData.java]: Ohh that's right! Java...
Didn't 👀 the `private` constructor. Sorry for the confusion!

File: libtelemetry/src/test/java/com/mapbox/android/telemetry/LocationEventTest.java [hunk 14]
```diff
@@ -1,21 +1,35 @@
...
 import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 public class LocationEventTest {
 
   @Test
   public void checksLocationEvent() throws Exception {
+    Context mockedContext = mock(Context.class);
```
Comment (Guardiola31337) [inline libtelemetry/src/test/java/com/mapbox/android/telemetry/LocationEventTest.java]: This block of code is duplicated 👀 `checksLocationType` test. Could you extract it into a `private` "factory" method called for example `obtainMockedContext`?

File: libtelemetry/src/test/java/com/mapbox/android/telemetry/LocationMapperTest.java [hunk 15]
```diff
@@ -1,16 +1,28 @@
...
 
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
 public class LocationMapperTest {
 
+  @Before
```
Comment (Guardiola31337) [inline libtelemetry/src/test/java/com/mapbox/android/telemetry/LocationMapperTest.java]: In general, we should avoid inline setup 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=65
Could you extract `test` code into a `private` "factory" method called for example `obtainMockedContext` and use it only in the tests that actually need it?
IMO, it reflects better what's going on in the test and you don't break the [triple A](https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40) "rule".

File: libtelemetry/src/test/java/com/mapbox/android/telemetry/SerializerTest.java [hunk 16]
```diff
@@ -1,26 +1,46 @@
...
 import java.util.Date;
 
 import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 public class SerializerTest {
 
+  @Before
```
Comment (Guardiola31337) [inline libtelemetry/src/test/java/com/mapbox/android/telemetry/SerializerTest.java]: In general, we should avoid inline setup 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=65
Could you extract `setupMapboxTelemetry` code into a `private` "factory" method called for example `obtainMockedContext` and use it only in the tests that actually need it?
IMO, it reflects better what's going on in the test and you don't break the [triple A](https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40) "rule".

File: libtelemetry/src/test/java/com/mapbox/android/telemetry/TelemetryClientNavigationEventsTest.java [hunk 17]
```diff
@@ -103,10 +108,20 @@ public GsonBuilder configure(GsonBuilder gsonBuilder) {
       }
     };
 
-  @Test
-  public void sendsTheCorrectBodyPostingAppUserTurnstileEvent() throws Exception {
+  @Before
```
Comment (Guardiola31337) [inline libtelemetry/src/test/java/com/mapbox/android/telemetry/TelemetryClientNavigationEventsTest.java]: In general, we should avoid inline setup 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=65
Could you extract `setupMapboxTelemetry` code into a `private` "factory" method called for example `obtainMockedContext` and use it only in the tests that actually need it?
IMO, it reflects better what's going on in the test and you don't break the [triple A](https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40) "rule".

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/BluetoothAudioType.java [hunk 18]
```diff
@@ -14,7 +14,9 @@ public void nextChain(AudioTypeResolver chain) {
 
   @Override
   public String obtainAudioType(Context context) {
-    AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
+    AudioManager audioManager = (AudioManager) MapboxTelemetry.applicationContext
+      .getSystemService(Context.AUDIO_SERVICE);
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/BluetoothAudioType.java]: Check formatting

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/HeadphonesAudioType.java [hunk 19]
```diff
@@ -14,7 +14,9 @@ public void nextChain(AudioTypeResolver chain) {
 
   @Override
   public String obtainAudioType(Context context) {
-    AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
+    AudioManager audioManager = (AudioManager) MapboxTelemetry.applicationContext
+      .getSystemService(Context.AUDIO_SERVICE);
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/HeadphonesAudioType.java]: Check formatting

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationUtils.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+import android.provider.Settings;
+
+class NavigationUtils {
+  private static final double PERCENT_NORMALIZER = 100.0;
+  private static final double SCREEN_BRIGHTNESS_MAX = 255.0;
+  private static final int BRIGHTNESS_EXCEPTION_VALUE = -1;
+
+  static int obtainVolumeLevel() {
+    AudioManager audioManager = (AudioManager) MapboxTelemetry.applicationContext
+      .getSystemService(Context.AUDIO_SERVICE);
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/NavigationUtils.java]: Check formatting

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/SpeakerAudioType.java [hunk 21]
```diff
@@ -14,7 +14,9 @@ public void nextChain(AudioTypeResolver chain) {
 
   @Override
   public String obtainAudioType(Context context) {
-    AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
+    AudioManager audioManager = (AudioManager) MapboxTelemetry.applicationContext
+      .getSystemService(Context.AUDIO_SERVICE);
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/SpeakerAudioType.java]: Check formatting

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java [hunk 22]
```diff
@@ -21,6 +29,119 @@
...
+        buffer.writeUtf8CodePoint(c > '\u001f' && c < '\u007f' ? c : '?');
+      }
+      return buffer.readUtf8();
+    }
+    return s;
+  }
+
+  static String obtainApplicationState() {
+    ActivityManager activityManager = (ActivityManager) MapboxTelemetry.applicationContext
+      .getSystemService(Context.ACTIVITY_SERVICE);
```
Comment (Guardiola31337) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryUtils.java]: Check formatting

File: libtelemetry/src/test/java/com/mapbox/android/telemetry/TelemetryClientNavigationEventsTest.java [hunk 23]
```diff
@@ -276,14 +285,11 @@ private Event obtainFeedbackEvent() {
     Date aDate = new Date();
     NavigationState navigationState = obtainDefaultNavigationState(aDate);
     FeedbackEventData navigationFeedbackData = new FeedbackEventData("anyUserId", "general",
-      "unknown", "audio");
+      "unknown");
```
Comment (Guardiola31337) [inline libtelemetry/src/test/java/com/mapbox/android/telemetry/TelemetryClientNavigationEventsTest.java]: Check formatting

**Pull request merged**